interfaces/cython/cantera/yaml2ck.py

```diff
@@ -0,0 +1,956 @@
+    skip_transport=False,
+):
+    """Writes Cantera solution object to Chemkin-format file.
+    Parameters
+    ----------
+    solution : cantera.Solution
+        Model to be written
+    output_path : path or str, optional
+        Path of file to be written; if not provided, use cd / 'solution.name'
+    reorder_reactions : bool, optional
+        Flag to reorder reactions based upon molecular weight
+    skip_thermo : bool, optional
+        Flag to skip writing thermo data
+    separate_thermo_file : bool, optional
+        Flag to write thermo data in a seperate file from mechanism
+    skip_transport : bool, optional
+        Flag to skip writing transport data in separate file
+    Returns
+    -------
+    output_file_name : list
+        List of paths to output mechanism files (.ck, .therm, .tran)
+    Examples
+    --------
+    >>> gas = cantera.Solution('gri30.cti')
+    >>> yaml2ck.write(gas)
+    [output_path / reduced_gri30.ck, output_path / reduced_gri30.therm,
+     output_path / reduced_gri30.tran]
+    """
+    if output_path:
+        if not isinstance(output_path, PurePath):
+            try:  # try to turn full path into pathlib Path
+                output_path = Path(output_path)
+            except:  # if previous fails, assume str is file name
+                output_path = Path.cwd() / output_path
+    else:
+        output_path = Path.cwd() / f"{solution.name}.ck"
+
+    if output_path.is_dir():
+        output_path = output_path / f"{solution.name}.ck"
+
+    if output_path.is_file():
+        output_path.unlink()
+
+    main_path = output_path.parents[0]
+    basename = output_path.stem
+    output_files = [output_path]
+
+    with open(output_path, "w") as mech_file:
+        # Write title block to file
+        if "description" in solution.input_header:
+            note = solution.input_header["description"]
+            mech_file.write(f"! {note}\n!\n")
+        else:
+            note = ""
+
+        if "! Chemkin file converted from Cantera solution object" not in note:
+            mech_file.write(
+                "! Chemkin file converted from Cantera solution object\n! \n\n"
+            )
+
+        # write species and element lists to file
+        element_names = "  ".join(solution.element_names)
+        mech_file.write("ELEMENTS\n" + f"{element_names}\n" + "END\n\n\n")
+
+        mech_file.write(species_data_text(solution.species()))
+
+        # Write thermo to file
+        if not skip_thermo and not separate_thermo_file:
+            mech_file.write(thermo_data_text(solution.species(), input_type="included"))
+
+        # Write reactions to file
+        max_rxn_width = 3 + max(
+            [len(rxn.equation) for rxn in solution.reactions()] + [48]
+        )
+        mech_file.write("REACTIONS  CAL/MOLE  MOLES\n")
+
+        # Write data for each reaction in the Solution Object
+        for n, reaction in enumerate(solution.reactions()):
+            reaction_string = ""
```

Comment:
Rather than doing `+=` on a string, it's typically faster to build up a list and then to `"\n".join()` at the end. This is because `+=` on a string requires allocating a whole new string and copying all the memory over (taking exponentially more time each time), whereas appending to a list is constant time.

Reply:
I did not know this. Thanks!